astropy/units/core.py

```diff
@@ -2313,6 +2318,8 @@ def __init__(
         # kwarg `_error_check` is False, the error checking is turned
         # off.
         if _error_check:
+            if scale == 0:
+                raise UnitScaleError("cannot create a unit with a scale of 0.")
```

Comment:
Wouldn't it be better if the code here called https://github.com/astropy/astropy/blob/32a6b7582f3029a65adc960f605d33663652211f/astropy/units/utils.py#L195
and it would be the job of that function to check if `scale` is zero? There's a `sanitize_scale()` call later on that we wouldn't need anymore.

Reply:
I like the idea, but I don't think we can get rid of the later calls: both in the short-cut in `__init__` and inside `_expand_and_gather`, the scale is multiplied by other scales raised to some power, which makes it possible that one gets, e.g., a result very close to unity that should be passed through `_sanitize_scale`. That said, in those places, some of the tests on the scale are not needed (e.g., whether it has a `.dtype`). So, I think you are right that this can be refactored, with an early check on numpy scalars and value of zero, and then the checks for near-unity still in the same place. How about I do that in follow-up? I think it would go beyond what I would be comfortable backporting.

Reply:
What you've done is a minimal bugfix, and the practice of the project is indeed not to backport refactoring pull requests.

Reply:
Do you want me to have a look at the refactoring or would you prefer to do it yourself?

Reply:
If you're interested, then please go ahead, thanks!